File: app/src/main/java/com/bijoysingh/quicknote/formats/Format.java
package com.bijoysingh.quicknote.formats;

import android.text.Spannable;
import android.text.SpannableString;
import android.util.Log;

import com.bijoysingh.quicknote.R;
import com.bijoysingh.quicknote.recyclerview.FormatListViewHolder;
import com.bijoysingh.quicknote.recyclerview.FormatTextViewHolder;
import com.github.bijoysingh.starter.recyclerview.MultiRecyclerViewControllerItem;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class Format {

  public static final String KEY_NOTE = "note";

  public FormatType formatType;

  public int uid;

  public String text;

  public Format() {
  }

  public Format(FormatType formatType) {
    this.formatType = formatType;
  }

  public Format(FormatType formatType, String text) {
    this.formatType = formatType;
    this.text = text;
  }

  public JSONObject toJson() {
    if (text == null || text.trim().isEmpty()) {
      return null;
    }

    Map<String, Object> map = new HashMap<>();
    map.put("format", formatType.name());
    map.put("text", text);
    return new JSONObject(map);
  }

  public static Format fromJson(JSONObject json) throws JSONException {
    Format format = new Format();
    format.formatType = FormatType.valueOf(json.getString("format"));
    format.text = json.getString("text");
    return format;
  }

  public static String getNote(List<Format> formats) {
    JSONArray array = new JSONArray();
    for (Format format : formats) {
      JSONObject json = format.toJson();
      if (json != null) array.put(json);
    }

    Map<String, Object> cache = new HashMap<>();
    cache.put(KEY_NOTE, array);
    return new JSONObject(cache).toString();
  }

  private int[] findFirstTwoOccurance(String source, String pattern) {
    Matcher matcher = Pattern.compile("(?=(aa))").matcher(source);
    List<Integer> position = new ArrayList<>();
    while (matcher.find()) {
      position.add(matcher.start());
    }

    if (position.size() < 2) {
      return null;
    }
    return new int[]{position.get(0), position.get(1)};
  }

  public static List<Format> getFormats(String note) {
    List<Format> formats = new ArrayList<>();
    try {
      JSONObject json = new JSONObject(note);
      JSONArray array = json.getJSONArray(KEY_NOTE);
      for (int index = 0; index < array.length(); index++) {
        try {
          Format format = fromJson(array.getJSONObject(index));
          format.uid = formats.size();
          formats.add(format);
        } catch (JSONException innerException) {
          Log.d(Format.class.getSimpleName(), innerException.toString());
        }
      }
    } catch (Exception exception) {
      Log.d(Format.class.getSimpleName(), exception.toString());
    }
    return formats;
  }

  public static List<MultiRecyclerViewControllerItem<Format>> getList() {
    List<MultiRecyclerViewControllerItem<Format>> list = new ArrayList<>();
    list.add(
        new MultiRecyclerViewControllerItem.Builder<Format>()
            .viewType(FormatType.MARKDOWN.ordinal())
            .layoutFile(R.layout.item_format_text)
            .holderClass(FormatTextViewHolder.class)
            .build());
    list.add(
        new MultiRecyclerViewControllerItem.Builder<Format>()
            .viewType(FormatType.TEXT.ordinal())
            .layoutFile(R.layout.item_format_text)
            .holderClass(FormatTextViewHolder.class)
            .build());
    list.add(
        new MultiRecyclerViewControllerItem.Builder<Format>()
            .viewType(FormatType.HEADING.ordinal())
            .layoutFile(R.layout.item_format_heading)
            .holderClass(FormatTextViewHolder.class)
            .build());
    list.add(
        new MultiRecyclerViewControllerItem.Builder<Format>()
            .viewType(FormatType.SUB_HEADING.ordinal())
            .layoutFile(R.layout.item_format_sub_heading)
            .holderClass(FormatTextViewHolder.class)
            .build());
    list.add(
        new MultiRecyclerViewControllerItem.Builder<Format>()
            .viewType(FormatType.QUOTE.ordinal())
            .layoutFile(R.layout.item_format_quote)
            .holderClass(FormatTextViewHolder.class)
            .build());
    list.add(
        new MultiRecyclerViewControllerItem.Builder<Format>()
            .viewType(FormatType.CODE.ordinal())
            .layoutFile(R.layout.item_format_code)
            .holderClass(FormatTextViewHolder.class)
            .build());
    list.add(
        new MultiRecyclerViewControllerItem.Builder<Format>()
            .viewType(FormatType.CHECKLIST_CHECKED.ordinal())
            .layoutFile(R.layout.item_format_list)
            .holderClass(FormatListViewHolder.class)
            .build());
    list.add(
        new MultiRecyclerViewControllerItem.Builder<Format>()
            .viewType(FormatType.CHECKLIST_UNCHECKED.ordinal())
            .layoutFile(R.layout.item_format_list)
            .holderClass(FormatListViewHolder.class)
            .build());
    return list;
  }

  public static FormatType getNextFormatType(FormatType type) {
    switch (type) {
      case BULLET_LIST:
        return FormatType.BULLET_LIST;
      case NUMBERED_LIST:
        return FormatType.NUMBERED_LIST;
      case HEADING:
        return FormatType.SUB_HEADING;
      case CHECKLIST_CHECKED:
      case CHECKLIST_UNCHECKED:
        return FormatType.CHECKLIST_UNCHECKED;
      case IMAGE:
      case SUB_HEADING:
      case CODE:
      case QUOTE:
      case TEXT:
      default:
        return FormatType.TEXT;
    }
  }
}

File: app/src/main/java/com/bijoysingh/quicknote/formats/Format.kt
package com.bijoysingh.quicknote.formats

import com.bijoysingh.quicknote.R
import com.bijoysingh.quicknote.recyclerview.EmptyFormatHolder
import com.bijoysingh.quicknote.recyclerview.FormatImageViewHolder
import com.bijoysingh.quicknote.recyclerview.FormatListViewHolder
import com.bijoysingh.quicknote.recyclerview.FormatTextViewHolder
import com.github.bijoysingh.starter.recyclerview.MultiRecyclerViewControllerItem
import org.json.JSONArray
import org.json.JSONException
import org.json.JSONObject
import java.util.*

class Format : Comparable<Format> {

  var formatType: FormatType = FormatType.TEXT

  var uid: Int = 0

  var text: String = ""

  var forcedMarkdown = false

  var metaData = HashMap<String, Any>()

  val markdownText: String
    get() {
      when (formatType) {
        FormatType.BULLET_LIST, FormatType.NUMBERED_LIST -> return "- " + text
        FormatType.HEADING -> return "# " + text
        FormatType.CHECKLIST_CHECKED -> return "\u2612 " + text
        FormatType.CHECKLIST_UNCHECKED -> return "\u2610 " + text
        FormatType.SUB_HEADING -> return "### " + text
        FormatType.CODE -> return "```\n$text\n```"
        FormatType.QUOTE -> return "> " + text
        FormatType.TEXT -> return text
        FormatType.IMAGE -> return ""
        else -> return text
      }
    }

  constructor() {}

  constructor(formatType: FormatType) {
    this.formatType = formatType
  }

  constructor(formatType: FormatType, text: String) {
    this.formatType = formatType
    this.text = text

    if (formatType === FormatType.TAG) {
      forcedMarkdown = true
    }
  }

  fun meta(key: String, defaultValue: Any): Any {
    return metaData[key] ?: defaultValue
  }

  fun addOrUpdateMeta(key: String, value: Any) {
    metaData[key] = value
  }

  fun toJson(): JSONObject? {
    if (text.trim { it <= ' ' }.isEmpty()) {
      return null
    }

    val map = HashMap<String, Any>()
    map["format"] = formatType.name
    map["text"] = text
    map["meta"] = metaData
    return JSONObject(map)
  }

  override fun compareTo(other: Format): Int {
    return when {
      other.formatType == FormatType.CHECKLIST_CHECKED && formatType == FormatType.CHECKLIST_UNCHECKED -> -1
      other.formatType == FormatType.CHECKLIST_UNCHECKED && formatType == FormatType.CHECKLIST_CHECKED -> 1
      else -> 0
    }
  }

  companion object {
    const val KEY_NOTE = "note"

    @Throws(JSONException::class)
    fun fromJson(json: JSONObject): Format {
      val format = Format()
      format.formatType = FormatType.valueOf(json.getString("format"))
      format.text = json.getString("text")

      if (json.has("meta")) {
        val metaJSON = json.getJSONObject("meta")
        for (key in metaJSON.keys()) {
          format.metaData.put(key, metaJSON.get(key))
        }
      }
      return format
    }

    fun getNote(formats: List<Format>): String {
      val array = JSONArray()
      for (format in formats) {
        val json = format.toJson()
        if (json != null) array.put(json)
      }

      val cache = HashMap<String, Any>()
      cache[KEY_NOTE] = array
      return JSONObject(cache).toString()
    }

    fun getFormats(note: String): List<Format> {
      val formats = ArrayList<Format>()
      try {
        val json = JSONObject(note)
        val array = json.getJSONArray(KEY_NOTE)
        for (index in 0 until array.length()) {
          try {
            val format = fromJson(array.getJSONObject(index))
            format.uid = formats.size
            formats.add(format)
          } catch (innerException: JSONException) {
          }
        }
      } catch (exception: Exception) {
      }
      return formats
    }

    val list: List<MultiRecyclerViewControllerItem<Format>>
      get() {
        val list = ArrayList<MultiRecyclerViewControllerItem<Format>>()
        list.add(
            MultiRecyclerViewControllerItem.Builder<Format>()
                .viewType(FormatType.TAG.ordinal)
                .layoutFile(R.layout.item_format_tag)
                .holderClass(FormatTextViewHolder::class.java)
                .build())
        list.add(
            MultiRecyclerViewControllerItem.Builder<Format>()
                .viewType(FormatType.TEXT.ordinal)
                .layoutFile(R.layout.item_format_text)
                .holderClass(FormatTextViewHolder::class.java)
                .build())
        list.add(
            MultiRecyclerViewControllerItem.Builder<Format>()
                .viewType(FormatType.HEADING.ordinal)
                .layoutFile(R.layout.item_format_heading)
                .holderClass(FormatTextViewHolder::class.java)
                .build())
        list.add(
            MultiRecyclerViewControllerItem.Builder<Format>()
                .viewType(FormatType.SUB_HEADING.ordinal)
                .layoutFile(R.layout.item_format_sub_heading)
                .holderClass(FormatTextViewHolder::class.java)
                .build())
        list.add(
            MultiRecyclerViewControllerItem.Builder<Format>()
                .viewType(FormatType.QUOTE.ordinal)
                .layoutFile(R.layout.item_format_quote)
                .holderClass(FormatTextViewHolder::class.java)
                .build())
        list.add(
            MultiRecyclerViewControllerItem.Builder<Format>()
                .viewType(FormatType.CODE.ordinal)
                .layoutFile(R.layout.item_format_code)
                .holderClass(FormatTextViewHolder::class.java)
                .build())
        list.add(
            MultiRecyclerViewControllerItem.Builder<Format>()
                .viewType(FormatType.CHECKLIST_CHECKED.ordinal)
                .layoutFile(R.layout.item_format_list)
                .holderClass(FormatListViewHolder::class.java)
                .build())
        list.add(
            MultiRecyclerViewControllerItem.Builder<Format>()
                .viewType(FormatType.CHECKLIST_UNCHECKED.ordinal)
                .layoutFile(R.layout.item_format_list)
                .holderClass(FormatListViewHolder::class.java)
                .build())
        list.add(
            MultiRecyclerViewControllerItem.Builder<Format>()
                .viewType(FormatType.IMAGE.ordinal)
                .layoutFile(R.layout.item_format_image)
                .holderClass(FormatImageViewHolder::class.java)
                .build())
        list.add(
            MultiRecyclerViewControllerItem.Builder<Format>()
                .viewType(FormatType.EMPTY.ordinal)
                .layoutFile(R.layout.item_format_fab_space)
                .holderClass(EmptyFormatHolder::class.java)
                .build())
        return list
      }

    fun getNextFormatType(type: FormatType): FormatType {
      when (type) {
        FormatType.BULLET_LIST -> return FormatType.BULLET_LIST
        FormatType.NUMBERED_LIST -> return FormatType.NUMBERED_LIST
        FormatType.HEADING -> return FormatType.SUB_HEADING
        FormatType.CHECKLIST_CHECKED, FormatType.CHECKLIST_UNCHECKED -> return FormatType.CHECKLIST_UNCHECKED
        FormatType.IMAGE, FormatType.SUB_HEADING, FormatType.CODE, FormatType.QUOTE, FormatType.TEXT -> return FormatType.TEXT
        else -> return FormatType.TEXT
      }
    }
  }
}
